### ISSUE TITLE
terrain function leaving holes

### ISSUE THREAD

**Issue by r2j2ritson** [rspatial/terra#452]
Hello, I'm not sure where to ask this, but I'm experiencing weird behavior with the `terrain` function. I have a 10 meter DEM for the entire state of Idaho and am trying to use terra::terrain() to calculate slope and aspect in radians (to be used for Sappington et al.'s (2007) Vector Ruggedness Measure). However, the output rasters each have NA values that seem to follow the gridlines of the merged tiles (see photos below). No such values are in the original DEM raster. Am I doing something wrong, or is this behavior a result of trying to calculate over such  a large area? Thanks in advance!

`elev <- terra::rast(paste0(filepath,"/DEM10M.tif")`
`sout <- paste0(filepath,"/sloperad.tif")`
`aout <- paste0(filepath,"/aspectrad.tif")`
`slp <- terra::terrain(elev, v="slope", unit="radians", neighbors = 8, filename= sout)`
`asp <- terra::terrain(elev, v="aspect", unit="radians", neighbors = 8, filename= aout)`

## DEM
`terra::plot(elev)`
![elev](https://user-images.githubusercontent.com/81387018/147131989-e1fb615a-9c14-45da-8767-fbf5e9c9ea97.PNG)

## Slope (note arrows)
`terra::plot(slp)`
![slp](https://user-images.githubusercontent.com/81387018/147131983-14fdf5f0-8cef-422b-955a-d33ea4562d9b.PNG)

## Aspect (note arrows)
`terra::plot(asp)`
![asp](https://user-images.githubusercontent.com/81387018/147131986-a21139db-e886-45be-ad52-9b0e418d273e.PNG)

**Comment by r2j2ritson** [comment]
Not sure if this related to #303 ??

**Comment by rhijmans** [comment]
Thank you, I can reproduce this with 

```
library(terra)
f <- system.file("ex/elev.tif", package="terra")
r <- rast(f)
x <- terrain(r, "slope", steps=3)
plot(x)
```

![slope](https://user-images.githubusercontent.com/11481397/147135003-f8678f3d-2502-4e28-b0d9-08f86e873e76.png)

So the problem is at the borders between chunks of data. That is why you normally only notice it with a large raster as in your case. I will fix this later today.

**Comment by r2j2ritson** [comment]
Thanks @rhijmans! Great job on this package by the way!

**Comment by rhijmans** [comment]
I believe this works now 

```
library(terra)
#terra 1.5.6
f <- system.file("ex/elev.tif", package="terra")
r <- rast(f)
v <- c("TPI", "TRI", "aspect", "flowdir", "slope", "roughness")
x <- terrain(r, v)
y <- terrain(r, v, steps=3)
                                                  
plot(y)
```
![terrain](https://user-images.githubusercontent.com/11481397/147156809-c97f5199-7eb2-41b7-ae84-b0c20be99616.png)

Only the results for "flowdir" are a bit different when running this in chunks, but that is, I think, because there is a random choice if there are ties. (you can use `set.seed` for each case, but it does not work across cases right now) 

```
all.equal(x, y)
#[1] "Component “flowdir”: Mean relative difference: 1.179328"
 
apply(values(x) == values(y), 2, mean, na.rm=T)
#      TPI       TRI    aspect   flowdir     slope roughness 
#1.0000000 1.0000000 1.0000000 0.9958767 1.0000000 1.0000000 
```

Thank you again for your input and for your encouraging words.

**Comment by r2j2ritson** [comment]
Thanks!

**Comment by ailich** [comment]
@rhijmans, is the `steps` argument available in all functions or just `terrain` as it seems to be useful for checking that out of memory data will be stitched together correctly without having to actually do a lengthy test on data that is too large to fit in memory?

**Comment by rhijmans** [comment]
Yes, you can use it any function. You can pass it directly to functions that have ellipses that take writing options. In all cases you can use `wopt=list(steps=2)` . And indeed it is there for testing purposes, possibly in combination with `SpatOptions("todisk=TRUE")`

**Comment by ailich** [comment]
Thanks!
